### PR TITLE
feat(runners): allow to use a shared encrypted AMI

### PR DESCRIPTION
**Pull request by v-rosa**
Currently its not possible to use an AMI encrypted using an external KMS key.

Related to https://github.com/philips-labs/terraform-aws-github-runner/issues/2927

Co-authored with @rsmolinski

### CODE REVIEW

File: modules/runners/main.tf [hunk 1]
```diff
@@ -38,7 +37,8 @@ locals {
     "linux"   = "${path.module}/templates/start-runner.sh"
   }
 
-  ami_filter = coalesce(var.ami_filter, local.default_ami[var.runner_os])
+  ami_kms_key_arn = var.ami_kms_key_arn != null ? var.ami_kms_key_arn : ""
```
Comment (npalm) [inline modules/runners/main.tf:40]: Can you add the variable also to the mult-runner module?

File: modules/runners/scale-up.tf [hunk 2]
```diff
@@ -88,6 +88,7 @@ resource "aws_iam_role_policy" "scale_up" {
     github_app_id_arn         = var.github_app_parameters.id.arn
     github_app_key_base64_arn = var.github_app_parameters.key_base64.arn
     kms_key_arn               = local.kms_key_arn
+    ami_kms_key_arn           = local.ami_kms_key_arn
```
Comment (npalm) [inline modules/runners/scale-up.tf:91]: The submodule pool can also create instances, please can you hadd the ami_kms key ass well to the pool module.

File: modules/runners/policies/lambda-scale-up.json [hunk 3]
```diff
@@ -52,6 +52,17 @@
...
             ],
             "Resource": "${kms_key_arn}"
+%{ endif ~}
+%{ if ami_kms_key_arn != "" ~}
+        },
+        {
+            "Effect": "Allow",
+            "Action": [
+                "kms:DescribeKey",
+                "kms:ReEncrypt*",
```
Comment (npalm) [inline modules/runners/policies/lambda-scale-up.json:62]: Is ReEncrypt needed?

File: modules/runners/policies/lambda-scale-down.json [hunk 4]
```diff
@@ -56,6 +56,17 @@
         "kms:Decrypt"
       ],
       "Resource": "${kms_key_arn}"
+%{ endif ~}
+%{ if ami_kms_key_arn != "" ~}
+    },
+    {
+      "Effect": "Allow",
```
Comment (npalm) [inline modules/runners/policies/lambda-scale-down.json:63]: Bit wondering why scale-down needs to be-able to decrypt en reencrypt. This lambda only terminates and describes instances.

**Review by npalm — CHANGES_REQUESTED** [review]
@v-rosa thanks for you work. In general looks good. Added a few comments.

File: modules/runners/policies/lambda-scale-up.json (hunk 3 above)
Comment (v-rosa) [reply modules/runners/policies/lambda-scale-up.json:62]: From my tests, yes. Here you can find a redacted version of the `ReEncrypt` event in CloudTrail triggered by the user name `<runnerprefix>-scale-up`

```
{
    "eventVersion": "1.08",
    "userIdentity": {
        "type": "AssumedRole",
        "principalId": "redacted:<runnerprefix>-scale-up",
        "arn": "arn:aws:sts::redacted:assumed-role/<runnerprefix>-action-scale-up-lambda-role/<runnerprefix>-scale-up",
        "accountId": "redacted",
        "accessKeyId": "redacted",
        "sessionContext": {
            "sessionIssuer": {
                "type": "Role",
                "principalId": "redacted",
                "arn": "arn:aws:iam::redacted:role/<runnerprefix>/<runnerprefix>-action-scale-up-lambda-role",
                "accountId": "redacted",
                "userName": "<runnerprefix>-action-scale-up-lambda-role"
            },
            "webIdFederationData": {},
            "attributes": {
                "creationDate": "2023-02-17T11:57:30Z",
                "mfaAuthenticated": "false"
            }
        },
        "invokedBy": "ec2-frontend-api.amazonaws.com"
    },
    "eventTime": "2023-02-17T13:45:01Z",
    "eventSource": "kms.amazonaws.com",
    "eventName": "ReEncrypt",
    "awsRegion": "us-east-1",
    "sourceIPAddress": "ec2-frontend-api.amazonaws.com",
    "userAgent": "ec2-frontend-api.amazonaws.com",
    "requestParameters": {
        "destinationEncryptionAlgorithm": "SYMMETRIC_DEFAULT",
        "sourceAAD": "redacted",
        "sourceEncryptionContext": {
            "aws:ebs:id": "redacted"
        },
        "destinationEncryptionContext": {
            "aws:ebs:id": "redacted"
        },
        "sourceEncryptionAlgorithm": "SYMMETRIC_DEFAULT",
        "destinationKeyId": "alias/aws/ebs",
        "destinationAAD": "redacted"
    },
    "responseElements": null,
    "requestID": "x",
    "eventID": "y",
    "readOnly": true,
    "resources": [
        {
            "accountId": "redacted",
            "type": "AWS::KMS::Key",
            "ARN": "arn:aws:kms:us-east-1:redacted:key/redacted"
        },
        {
            "accountId": "redacted",
            "type": "AWS::KMS::Key",
            "ARN": "arn:aws:kms:us-east-1:redacted:key/redacted"
        }
    ],
    "eventType": "AwsApiCall",
    "managementEvent": true,
    "recipientAccountId": "redacted",
    "sharedEventID": "z",
    "eventCategory": "Management"
}
```

File: modules/runners/scale-up.tf (hunk 2 above)
Comment (v-rosa) [reply modules/runners/scale-up.tf:91]: sure!

File: modules/runners/main.tf (hunk 1 above)
Comment (v-rosa) [reply modules/runners/main.tf:40]: sure!

File: modules/runners/policies/lambda-scale-down.json (hunk 4 above)
Comment (v-rosa) [reply modules/runners/policies/lambda-scale-down.json:63]: Indeed this is not needed for scale down. Will remove.

File: modules/runners/main.tf (hunk 1 above)
Comment (v-rosa) [reply modules/runners/main.tf:40]: done.

File: modules/runners/scale-up.tf (hunk 2 above)
Comment (v-rosa) [reply modules/runners/scale-up.tf:91]: done

File: modules/runners/policies/lambda-scale-down.json (hunk 4 above)
Comment (v-rosa) [reply modules/runners/policies/lambda-scale-down.json]: done

**Review by npalm — APPROVED** [review]

**Comment by npalm** [comment]
@v-rosa @rsmolinski Thanks!

**Pull request merged**